storage/src/main/scala/coop/rchain/storage/regex/PathRegex.scala

```diff
@@ -65,10 +120,77 @@ private[regex] object PathToken {
 case class PathRegex(tokens: List[PathToken], options: RegexOptions) {
   def withOptions(newOptions: RegexOptions): PathRegex = new PathRegex(tokens, newOptions)
 
-  def toPath(args: List[(String, Any)] = Nil): Either[Throwable, String] =
-    throw new NotImplementedError("TODO")
+  /**
+    * Takes some arguments (argument can be a value or a sequence of values),
+    * and builds an Uri-path
+    */
+  def toPath(args: Map[String, Iterable[String]],
+             encode: String => String = PathRegex.encodeUriComponent): Either[Throwable, String] =
+    Try(tokens.map(_.formatSegment(args, encode)).mkString).toEither
 
-  def toPath(arg: (String, Any)): Either[Throwable, String] = toPath(arg :: Nil)
+  val keys: List[PathToken] = tokens.filter(_.isToken)
+
+  /**
+    * Returns Regex that is able to parse our tokens with defined options
+    */
+  lazy val Regex: Either[Throwable, Regex] = {
+    val endsWith = ("$" :: options.endsWith.map(PathRegex.escapeString)).mkString("|")
+
+    val route: List[Option[String]] = tokens.flatMap { token =>
+      if (token.isRawPathPart) {
+        token.rawPathPart.map(tokenRawPart => Some(PathRegex.escapeString(tokenRawPart)))
+      } else {
+        token.pattern.map(tokenPattern => {
+          val prefix = token.prefix.map(PathRegex.escapeString).getOrElse("")
+          val capture = if (token.repeat) {
+            s"(?:$tokenPattern)(?:$prefix(?:$tokenPattern))*"
+          } else {
+            tokenPattern
+          }
+
+          if (token.optional) {
+            if (token.partial) {
+              Some(s"$prefix($capture)?")
+            } else {
+              Some(s"(?:$prefix($capture))?")
+            }
+          } else {
+            Some(s"$prefix($capture)")
+          }
+        })
+      }
+    }
+
+    val routeFinish: List[String] = if (options.end) {
+      val strictFinish = if (options.strict) {
+        Nil
+      } else {
+        s"(?:${options.delimiter})?" :: Nil
```

Comment:
Same comment as last time regarding using `::`.

 I tend to prefer:

```scala
List(s"(?:${options.delimiter})?")
```

In the `Nil` cases, you can use `List.empty` if you want consistency (also a good thing IMHO)

Reply:
Sure. Updated everywhere and replaced Nil's as well.